### DISCUSSION TITLE
Wrong audio on multi language videos

### DISCUSSION THREAD

**Discussion by juleppar** [aclap-dev/video-downloadhelper#1507] [Help and Q&A]
### Your question

Trying to download a video on Firefox and it has multiple audio tracks for various languages (an AI feature?). My problem is prety much a duplicate of the one reported by Fuzzlix on September 26 except that for me the language picked by VDH mostly seems to be German. This is rapidly concerning more and more channels and it would really be helpful if we could choose not just the resolution but the audio stream as well? Below video as an example:
https://www.youtube.com/watch?v=M8tXHQMxEVQ

### Addon & CoApp version and details

version: 9.1.0.44
target: mozilla
channel: stable
lang: en-US
coapp: {"found":true,"path":"C:\\Program Files\\DownloadHelper CoApp\\vdhcoapp.exe","version":"2.0.19","new_version":false}
license: {"accepted":true,"email":"<removed>","key":"0f759f6d52d84536"}
platform: x86-64 win
UA: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:131.0) Gecko/20100101 Firefox/131.0
{
    "valid_license_message_has_been_displayed": true,
    "used_history_button": true,
    "download_directory": "G:\\independent\\VideoDownloadHelper",
    "last_download_directory": {
        "__serializer_tag": "option_some",
        "__serializer_value": {
            "__serializer_tag": "primitive"
        }
    },
    "theme": "dark",
    "view_options": {
        "all_tabs": false,
        "low_quality": false,
        "sort_by_status": true,
        "sort_reverse": false,
        "show_button_clean": true,
        "show_button_clean_all": false,
        "show_button_convert_local": false,
        "hide_downloaded": false
    },
    "use_sidebar": false,
    "open_count_store": 1839,
    "successfull_dl": 55,
    "never_show_successfull_dl_message": true
}

**Comment by dilotec-2015** [comment]
i am facing the same problem. please add the possibility to pick the wished language

**Reply by weeble21613** [reply]
I am also having this problem which started about 2 weeks back, and it seems like only certain youtubers i get this issue from. Not all of them.  I have my setting on Eng and its changing to Ger.
example below
https://www.youtube.com/watch?v=3I7pJxgOTAA&t=273s

everything was working fine up until 2 weeks back.

**Reply by GrampaWildWilly** [reply]
I do not work for VDH.  I'm just another user.  I've said that elsewhere in this thread (and throughout this forum) but I feel the need to repeat it once again.

Let's all face reality here.  Our dev does not look like he's jumping on this issue.  So you have a choice.  Either you wait an eternity for VDH to solve this problem.  Or you learn something new that will allow you to download this content right now.  I'm going to show you how I did it.  You can choose to learn this or not.  It's up to you.

First, I loaded up the YouTube page.  VDH recognized something.  But instead of just using VDH to download what we all agree is going to be in the wrong language, do this:

![#01](https://github.com/user-attachments/assets/9c6f9e93-e455-427c-a61a-c038b7f0a1db)
![#02](https://github.com/user-attachments/assets/3464a84b-16b7-48ce-b38f-d74e4ae68fbe)

That opens a new browser tab with a bunch of details.  The VDH menu will, annoyingly, still be open.  Just hit Escape to dismiss it.  Then . . .

![#03](https://github.com/user-attachments/assets/6bf96d70-23f4-4c3b-99a7-a0e22e739994)

That URL is for an HLS master manifest somewhere on googlevideo.com, which is YouTube.

Then create an invocation of ffprobe to analyze that manifest.  You have ffrpobe.  It's distributed as part of our CoApp.  You can find where it's located by looking in the VDH settings:

![#04](https://github.com/user-attachments/assets/061f5612-10f9-47f7-8f8c-9f0b5181bbcd)

When I ran that URL through ffprobe, I got these results:

[ffprobe.txt](https://github.com/user-attachments/files/18050743/ffprobe.txt)

The exact ffprobe command I executed was this:

```
"C:\DownloadHelper CoApp\ffprobe.exe"  -protocol_whitelist file,crypto,data,http,https,tls,tcp -strict experimental -analyzeduration 25000000 -probesize 250000000 -user_agent "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:133.0) Gecko/20100101 Firefox/133.0" -allowed_extensions ALL "https://manifest.googlevideo.com/api/manifest/hls_variant/expire/1733648143/ei/rwpVZ4O1A93EzN0Pjsui0QE/ip/2603%%3A6011%%3Ac300%%3Ae8c0%%3Ab43c%%3A3b35%%3Aaa03%%3Aa131/id/dc8ee927180e4c00/source/youtube/requiressl/yes/xpc/EgVo2aDSNQ%%3D%%3D/playback_host/rr1---sn-vgqsrns6.googlevideo.com/met/1733626543%%2C/mh/0r/mm/31%%2C26/mn/sn-vgqsrns6%%2Csn-p5qlsn6l/ms/au%%2Conr/mv/m/mvi/1/pl/32/rms/au%%2Cau/hfr/1/demuxed/1/tts_caps/1/maudio/1/initcwndbps/1407500/spc/qtApATLgBR8dliqyB6RV6JRoJvK5UIJL4d5MSgYRatnEAKusXg/vprv/1/go/1/rqh/5/mt/1733626124/fvip/1/nvgoi/1/short_key/1/ncsapi/1/keepalive/yes/fexp/51326932%%2C51335594%%2C51347746/dover/13/itag/0/playlist_type/DVR/sparams/expire%%2Cei%%2Cip%%2Cid%%2Csource%%2Crequiressl%%2Cxpc%%2Chfr%%2Cdemuxed%%2Ctts_caps%%2Cmaudio%%2Cspc%%2Cvprv%%2Cgo%%2Crqh%%2Citag%%2Cplaylist_type/sig/AJfQdSswRgIhAILV08htaX8hV1ZVZo5v6H74xmqnWZBUf3vjQUXStXM_AiEAwg9t8vEZ_UstrGaEj8AACVt5t7NVLF6ceNRtz7F7-Og%%3D/lsparams/playback_host%%2Cmet%%2Cmh%%2Cmm%%2Cmn%%2Cms%%2Cmv%%2Cmvi%%2Cpl%%2Crms%%2Cinitcwndbps/lsig/AGluJ3MwRgIhANqc0vXMmQTy26R8PRnvcgmSkvC8C_s7e9zcyGif3tOlAiEAjlDm_pElAzrodwAM7IZHCPEy3zhjJ3RpjEoHk_-H22c%%3D/file/index.m3u8" 2>>"Q:\To Watch\ffprobe.txt"
```

You'll notice that all the % signs appear as %%.  That's because I ran this from a .bat file, not as a raw command I typed into a command prompt window.  You want to be doing these things from .bat files, not typing them in every time.

Just study that ffprobe log file.  Scroll it to the bottom, then scroll up until you find the line that reads:

`Input #0, hls, from`

What follows that line is the important information provided by ffprobe.  This output is in this format:

```
  Program x 
  Stream #x:y
  Stream #x:y
  Stream #x:y
  Stream #x:y
  Program x 
  Stream #x:y
  Stream #x:y
  Stream #x:y
  Stream #x:y
  Program x 
  Stream #x:y
  Stream #x:y
  Stream #x:y
  Stream #x:y
```

Each Program is a variant as recognized by VDH.  The Streams within a Program constitute the tracks of the video: video, audio, sometimes captions, sometimes the useless Timed_ID3 data.  In this case, each Program begins with several Streams describing the multiple available audio tracks.  You can see the language of each audio track in the ffprobe report.  After all the audio tracks, you will find the Stream that describes the video track.  You will see the video resolution there.

In this ffprobe report, there are a couple of English audio tracks.  I pick English because that's my mother tongue.  If you want another language, just search for it.  The marker for English is `(en-US)`.  When you search on that string, you will find Streams #0:5 & #0:11, a few occurrences of each in different Programs.  The Stream identifiers are important.  You'll see below how they are used.

As an aside, you'll note in this ffprobe report that Stream #0:0, which is the first audio track listed, is marked `(de-DE)`.  That's German.  That's why VDH gave you a video with a German audio track.  VDH today does not have any smarts about the possibility that there could be multiple audio tracks.  It just picks the first one it finds & doesn't look any further.  This is the VDH bug that we all want corrected.

Now, look for your video track.  I'm going to get the `1920x1080` video so you search on that in the file.  That turns up Streams #0:18, #0:26, & #0:27.  The first video Stream shows `h264 (High)` on that line.  That indicates that it is in MP4 format.  The other 2 video Streams show `vp9 (Profile 0)`.  That indicates that they are in MKV format.  In order to choose the one I want, I look below the Stream line at the line that shows `variant_bitrate`.  The values for these 3 video Streams are, respectively, 5480335, 4493097, & 5823008.  The third one is the highest value.  That means it is the "best" quality video Stream at resolution 1920x1080.  So I will be getting Stream #0:27.  That is the video Stream within Program 15.  The English audio Stream within Program 15 is Stream #0:11.  So that is the one I'm going to get.

Now I concoct an invocation of ffmpeg to download the video with the English audio track.  Here is the log from that operation:

[YouTube Winter Video mkv Log.txt](https://github.com/user-attachments/files/18050746/YouTube.Winter.Video.mkv.Log.txt)

At the very top of the log you can see the exact ffmpeg command I executed.  Look carefully for the parameters `-map 0:27` & `-map 0:11`.  This is why you spend time analyzing an ffprobe report.  It informs your ffmpeg command.  You get the Stream numbers.  You also determine whether the output should be MP4 or MKV.

About 1 screenload up from the bottom of the log, there is a short report of the results of the download.  It shows the duration of the download, the size of the resulting file, & the download speed.  YouTube was cooperating tonight.  I got a great download speed.  YouTube is very inconsistent.  You can get garbage download speeds.  You can get wonderful download speeds, like I happened to get tonight.  You can get anything in between.  I'm not sure what determines the download speed.  Time of day?  Load on the YouTube servers?  The temperature outside?  The mood of the webmaster on duty?  I can't say.  You get what you get.  Tonight, I got lucky.  Mileage, as they say, varies.  Varies wildly.

At the very bottom of that log is a report from ffprobe on the resulting file on my system.  Note how my earlier ffprobe was against a URL for an HLS master manifest online.  This ffprobe is against an MKV video file on my system.  The ffprobe utility is versatile, not to mention critically important in this process.

Of course, the video played fine in VLC, with both video & English audio right to the end.  I didn't sit & watch it, just sampled it at intervals to make sure it was all there.

Now, like I said before, it's up to you whether you want to sit idly, waiting for somebody else to fix VDH.  Or you can take matters into your own hands right now & solve your own problem.  It's up to you.  You will likely find this tutorial on ffmpeg of great usefulness:

https://groups.google.com/g/video-downloadhelper-q-and-a/c/oYDFx367MSY

**Reply by weeble21613** [reply]
Thank you GrampaWildWilly

**Reply by pbackx** [reply]
Thanks @GrampaWildWilly 
I had a video where Video DownloadHelper kept downloading the lowest bitrate video, instead of the HD one. I was able to download the high quality one with your explanation.

**Comment by GrampaWildWilly** [comment]
I do not work for VDH.  I'm just another user.

You'll want to study #956 & #985.

**Reply by juleppar** [reply]
Thanks. 
I'm still hoping for something more automated. As I said this is moving through Youtube channels quite rapidly. I suspect this will concern all videos on the platform in very near future.

**Reply by GrampaWildWilly** [reply]
Yes, I think we all are.  We can only express our concerns here, point out samples with multiple language tracks, & hope the dev pays attention.  Maybe there is a day in the (probably distant) future when the feature we hope for materializes.

**Comment by hranier64** [comment]
While I agree with the others about the issue as it's frustrating, there may also be some variables that they need to figure out.  For example, Youtube may use randomly generated codes or segments within the URL to access different parts of the database to retrieve the language tracks for example (and these may possibly be updated on a daily basis) thus making it harder for programs like VDH to work properly.  So while it's a new feature for Youtube, they may have also seen it as an opportunity to make it harder for such programs to work.  The other thing is that not all videos may have the same number of audio tracks (some may have more languages than others) so this may also further complicate things (although part of me thinks that whatever codes are being used for US English for example, would remain the same from video to video regardless of the number of tracks available.

Regarding the copy of the URL, in the extension, there is an option to copy the video URL, so you don't necessarily need to go through the steps outlined above to get the URL, although I haven't verified if the URL that is copied is the same one shown above in the detailed explanation.

**Reply by GrampaWildWilly** [reply]
The URL shown in the Hit Details, which is what I show above in my explanation of how to do this with ffmpeg directly, is most definitely NOT the URL of the video.  The URL of the video has `watch?v=` in it.  The URL from the Hit Details points to an HLS master manifest, a completely different object.  It's a URL that contains `m3u8`.  The HLS master manifest does not show up in the Developer Tools/Network Monitor.  VDH is doing something to make YouTube generate an HLS master manifest.  It's a bit of magic that is not explained anywhere.  It is probably part of the trade secrets that justify the existence of a paid license for VDH.  The fact that I can use information supplied by VDH itself to accomplish this download says to me that VDH could be made to work on multi-language videos.  We can only hope that eventually, that's what happens.  It's just a matter of time.  But the developer's absence from this forum for more than 4 months now makes me think it will not be a feature we will get any time soon.

**Comment by ghmagnet** [comment]
Thank you GrampaWildWilly for such a comprehensive breakdown of the process.

I wrote a bash script that opens my youtube channels in tabs, then fetches the titles and checks them against a list in a file, only downloading any new videos it finds, adding completed downloads to the file. The switch to German language certainly caught me out on many channels I follow.

Using your explaination, I have rewritten a new version of the script that fetches all the information, extracts the stream ID's and now downloads the videos correctly in English.

Your efforts in solving this are truly appreciated here showing a path for me to follow.

Regards
magnet :)

**Reply by doncisco** [reply]
@ghmagnet, would you be willing to share your bash script? I'm struggling with the same issue and it would be nice to be able to automate some of the work. It would be greatly appreciated :)

**Reply by ghmagnet** [reply]
Unfortunately it's very specific to my system and simply would not function anywhere else and uses many commands not installed by default on the OS, plus it's not Windows but linux here. I'm just a hobbyist at programming so there must be many better ways to pull all the info together and get that ffmpeg command to run than what I've throw together.

Have a "bash" at programming something up from scratch. All I'm doing is saving the information from the video Details page to a file, using xclip -0 > filename.txt then using grep to search for the masterManifest line in that file I need to run ffprobe. I then basically repeat by searching ffprobe output file for the 1920x1080 line and getting the video stream number 0:18, then repeat again to get the audio stream number 0.11. It's basically doing the same stuff a few times.

Watching channels here is very hard with video stutter/buffering due to terrible internet speeds around here, so even a very slow download to file makes it possible to enjoy a video uninterupted.

With better internet I doubt I would have bothered and just watched them straight in the browser.

**Reply by doncisco** [reply]
@ghmagnet Thanks for the general steps - it helps.

**Comment by mouse-the** [comment]
Here is a quick script based on information in this thread. There seem to be a lot of edge cases, and things may change quickly, but it works for me right now. 

#!/bin/bash
#takes a url...
if [ -z "$1" ]; then
        echo "Please provide a URL for ffprobe..."
        exit 1
fi

useragent='Mozilla/5.0 (X11; Linux x86_64; rv:128.0) Gecko/20100101 Firefox/128.0'
outfile='./youtube.mkv'
split='[() :#]'
language='(en-US)'
aquality='(LC)'
vquality='vp9.*1280'

ffprobe  -protocol_whitelist http,https,tls,tcp -strict experimental -analyzeduration 25000000 -probesize 250000000\
 -user_agent "$useragent" -allowed_extensions ALL $1 2>./2.out

audio=$(awk -e "/$language.*$aquality/ { split(\$0, line, \"$split\");  print line[5] \":\" line[6]; exit }" 2.out)
if [ -z "$audio" ]; then
        audio=$(awk -e "/$aquality.*stereo/ { split(\$0, line, \"$split\");  print line[5] \":\" line[6]; exit }" 2.out)
        echo "could not find audio matching $language, using first stream matching $aquality..."
fi

video=$(awk -e "/$vquality/ { split(\$0, line, \"$split\");  print line[5] \":\" line[6]; exit }" 2.out)
if [ -z "$video" ]; then
        echo "could not find a video stream..."
        exit
fi

rm ./2.out
echo "Audio Stream $audio, Video Stream $video"

ffmpeg -fflags +discardcorrupt  -user_agent "$useragent"  -hwaccel auto -protocol_whitelist file,crypto,data,http,https,tls,tcp -strict experimental  -analyzeduration 25000000 -probesize 250000000  -allowed_extensions ALL  -i $1 -codec: copy  -map $video  -map $audio "$outfile" 2>./ffmpg.log

**Reply by doncisco** [reply]
@mouse-the thanks for the script. I'm going to try it out

**Reply by vperalta57** [reply]
How do I run the script on windows 10?

**Reply by mouse-the** [reply]
Although there are probably ways to run that bash script under windows, I think it would be better for someone to write a version for windows. I do not know if microsoft ships something like awk that can pull the stream numbers out of the ffprobe results in the same way. 

If awk is available, someone could hardcode the quality and language into the following program in order to avoid most of the quote escaping weirdness. The output of ffprobe goes into awk, which should spit out the 2 -map arguments that need to go into ffmpeg...

awk -e '
BEGIN {  FS = "[ #:(]+" }
/(en-US).*(LC)/ {  audio = ($3 ":" $4) }
/(LC).*stereo/ { afb = ($3 ":" $4) }
/vp9.*1280/ { video = ($3 ":" $4) }
/Video: vp9/ { vfb = ($3 ":" $4) }
END {
  if (audio == "") audio = afb
  if (video == "") video = vfb
  print ( "-map " video " -map " audio ) }
' ./2.out

It also seems likely that ffprobe can be convinced to only output the needed information, which could make this easier, but I failed.
Some people might want to just have ffmpeg collect all the audio streams and then let the player choose one. That might simplify the script...

**Comment by drouu** [comment]
number one frustration, right here. nothing like downloading a large video, only to find its in... i dunno, portuguese or arabic or something random.

love the energy of everyone posting their ffmpeg arcana, but if i wanted to do that i wouldn't be using a browser extension in the first place. and i lament anyone willing to run random scripts they find on the internet, probably with no comprehension of what they're doing. that's a quick way to make the cybersecurity segment of the evening news.
